# Project: Exploring Hacker News Posts
In this project, we'll work with a data set of submissions to popular technology site __[Hacker News](https://news.ycombinator.com/)__.

Hacker News is a site started by the startup incubator __[Y Combinator](https://www.ycombinator.com/)__, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.



# Dataset

We are going to use the data set "hacker_news.csv". You can find the data set __[here](https://www.kaggle.com/hacker-news/hacker-news-posts)__, but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

# Goal

In this small project, we're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question.

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?

Do posts created at a certain time receive more comments on average?

# Summary of Results

After analyzing the data, we can infer that: 
    
Show HN receives more comments than Ask HN on average

Base on the Show HN and Ask HN data, we think that late night US eastern time is the best time to post on Hacker News


# Exploring and Preparing the dataset

In [89]:
# open and read 'hacker_news.csv'
from csv import reader
open_filed = open('hacker_news.csv')
read_filed = reader(open_filed)
hn = list(read_filed)
headers = hn[0]
hn = hn[1:]
print("The headers of the dataset:")
print('\n')
print(headers)
print('\n')
# print the first 5 rows of the dataset
print('The first 5 rows of the dataset:')
print('\n')
for row in hn[:5]:
    print(row)
    print('\n')
    


The headers of the dataset:


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


The first 5 rows of the dataset:


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.h

Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.


In [23]:
# initiate a list contains Ask HN posts
ask_posts = []

# initiate a list contains Show HN posts
show_posts = []

# initiate a list contains all other posts
other_posts = []
# populate the lists
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

# checking the length of each list
print("Length of ask_posts, show_posts, other_posts respectively:")
print(len(ask_posts), len(show_posts), len(other_posts))
print('\n')
# first five rows in the ask_posts list of lists:
print('First five rows in the ask_posts list of lists:')
print('\n')
for item in ask_posts[:5]:
    print(item)
    print('\n')
    


Length of ask_posts, show_posts, other_posts respectively:
1744 1162 17194


First five rows in the ask_posts list of lists:
['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']


['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']


['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']


['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']


['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']




In [25]:
# first five rows in the show_posts list of lists:
print('First five rows in the show_posts list of lists:')
print('\n')
for item in show_posts[:5]:
    print(item)
    print('\n')

First five rows in the show_posts list of lists:


['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03']


['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']


['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05']


['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11']


['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']




# Do Ask HN or Show HN receive more comments on average?

Next, let's determine if Ask posts or Show posts receive more comments on average.

In [87]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments

avg_ask_comments = total_ask_comments/ len(ask_posts)
print('Average number of comments on ask posts: {:.2f}'.format(avg_ask_comments))

for row in show_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments

avg_show_comments = total_ask_comments/ len(ask_posts)
print('Average number of comments on show posts: {:.2f}'.format(avg_show_comments))


Average number of comments on ask posts: 14.04
Average number of comments on show posts: 20.91


As we can see on average Show HN posts received more comments than Ask HN posts

# Do posts created at a certain time receive more comments on average?

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. 

To accomplish this we will alculate the average number of comments for posts created during each hour of the day.



First let's look at the data foe Ask posts

In [92]:
# Calculate the average number of comments for posts created
# during each hour of the day, and extract the data into the result_list
import datetime as dt
result_list = []

for row in ask_posts:
    date = row[-1]
    num_comments = int(row[4])
    result_list.append([date, num_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    hour = date.hour
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
        
avg_by_hour = []
for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

# for row in avg_by_hour[:5]:
#     print(row)
#     print('\n')


In [91]:
# sort the result_list to find the answer
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
# for row in swap_avg_by_hour:
#     print(row)
#     print('\n')
    
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments: ")   
print('\n')

for row in sorted_swap[:5]:
    hour = dt.datetime.strptime(str(row[1]), "%H")
    hour = hour.strftime("%H:%M")
    print("{}: {:.2f} average comments per post".format(hour, row[0]))   
    print('\n')

Top 5 Hours for Ask Posts Comments: 


15:00: 38.59 average comments per post


02:00: 23.81 average comments per post


20:00: 21.52 average comments per post


16:00: 16.80 average comments per post


21:00: 16.01 average comments per post




For Ask Posts, we can see that the posts receive the most comments are those posted from 15:00 with around 38.59 comments per hour on average. Late night posts around 2:00 also recieve a fair bit of interactions with about 23.81 comments on average. Posts made at around 16:00 also recieve about 16.8 comments on average. Hence we can infer that:

the best time to post in the ask HN is in the middle of the afternoon from 15:00 to 16:00 or late at night at about 2:00 US eastern time



Now, let's look at the data foe Show posts

In [93]:
# Calculate the average number of comments for posts created
# during each hour of the day, and extract the data into the result_list
import datetime as dt
result_list = []

for row in show_posts:
    date = row[-1]
    num_comments = int(row[4])
    result_list.append([date, num_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    hour = date.hour
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
        
avg_by_hour = []
for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])


In [94]:
# sort the result_list to find the answer
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
# for row in swap_avg_by_hour:
#     print(row)
#     print('\n')
    
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Show Posts Comments: ")   
print('\n')

for row in sorted_swap[:5]:
    hour = dt.datetime.strptime(str(row[1]), "%H")
    hour = hour.strftime("%H:%M")
    print("{}: {:.2f} average comments per post".format(hour, row[0]))   
    print('\n')

Top 5 Hours for Show Posts Comments: 


18:00: 15.77 average comments per post


00:00: 15.71 average comments per post


14:00: 13.44 average comments per post


23:00: 12.42 average comments per post


22:00: 12.39 average comments per post




The two stand out times to make a Show Post are around 18:00 or 0:00. 

The posts at around 18:00 have around 15.77 comments on average. 

While the posts around 0:00 have around 15.71 comments on average.

# Conclusion


Base on the data on Show Posts and Ask Posts, we can infer that:

Show Posts tend to have more interactions than Ask Post on average.

Late night post, US eastern time, on Hacker News tend to get most comments, and thus late at night may be the best time to post on Hacker News